### DISCUSSION TITLE
Trying to configure an existing bucket

### DISCUSSION THREAD

**Discussion by AmitLustiger** [vladimir-bukhtoyarov/bucket4j#438] [Q&A]
Hey there!
I'm having a problem with overriding a bucket by key.

I use 2 API's in my code.

**The first** tries to consume a token from a bucket by key, and if there is no bucket exists for that key, a new Bucket is created with default Bandwidth values (hardcoded).

Supplier<BucketConfiguration> configSupplier = getConfigSupplier();
        return proxyManager.builder().build(key, configSupplier);

    private Supplier<BucketConfiguration> getConfigSupplier() {
      Refill refill = Refill.intervally(10, Duration.ofMinutes(1));
              Bandwidth limit = Bandwidth.classic(100, refill);
      
              return () -> (BucketConfiguration.builder()
                      .addLimit(limit)
                      .build()); 
  }

**The 2nd** allows to change the configuration of an existing bucket.

For example, if I created a bucket with the first API with the key "key1" and now I use the 2nd API to reconfigure and replace the bucket created with:

Bucket newBucket = proxyManager.builder().**withImplicitConfigurationReplacement**(1, TokensInheritanceStrategy.PROPORTIONALLY).
                    build(key, newBucketConfiguration);

where key = "key1"

When I try to use the first API again to tryConsume a token from the bucket, the bucket is still the original one. 
(I debugged it and saw that the Configuration is the same).

I saw in the documentation that we need to use withImplicitConfigurationReplacement in order to change it.

I also tried replacing the following line in the first API:

        return proxyManager.builder().build(key, configSupplier);
**To**
        proxyManager.builder().withImplicitConfigurationReplacement(2, TokensInheritanceStrategy.PROPORTIONALLY).
                            build(key, configSupplier);

**Also**, 
I saw that there is a solution using ** replaceConfiguration**.
But that method is supposed to be performed on an object of type Bucket.
How can I get my bucket using its key?
ProxyManager supplies only getProxyConfiguration method that returns BucketConfiguration object.

Thank you!

**Comment by vladimir-bukhtoyarov** [comment]
1. You should interract in any way with bucket after using *withImplicitConfigurationReplacement*, reconfiguration will not take effect without interaction, *getAvailableTokens* is suitable to touch the bucket.
```java
proxyManager.builder().withImplicitConfigurationReplacement(2, TokensInheritanceStrategy.PROPORTIONALLY).
build(key, configSupplier)
.getAvailableTokens();
```
2. If you use some expiration for buckets, then hardcoded configuration is not good solution in such case, because reconfiguration  esffect can be lost after expiration.
3. *replaceConfiguration* is more prefferable way for reconfiguration of particular bucket. 


> I saw that there is a solution using ** replaceConfiguration**.
But that method is supposed to be performed on an object of type Bucket.
How can I get my bucket using its key?
ProxyManager supplies only getProxyConfiguration method that returns BucketConfiguration object.

Just build bucket, and reconfigure it, bucket-proxy is lazy, there is no double interraction with storage.

**Reply by AmitLustiger** [reply]
First of all, thank you!
Let's focus on the 3rd answer- replaceConfiguration.

You said:
Just build bucket, and reconfigure it, bucket-proxy is lazy, there is no double interraction with storage.

But if I created a bucket, and after a long time I decide that I want to reconfigure it, how can I access it to do so?
Just building a new one and reconfiguring it does not work.
What is the way to do so?

This is what I tried to do and failed:

I have 2 methods:

```
public ResponseEntity<?> allowOrLimit(String storeId, String path) {
        Bucket bucket = rateLimitConfig.resolveBucket(storeId + "--" + path);
}

 public ResponseEntity<?> configureBucket(BucketConfigurationDTO bucketConfig) {
        ArrayList<Bucket> buckets = rateLimitConfig.resolveBucket(bucketConfig);
}
```

This is the config class:

```
private final ProxyManager<String> proxyManager;

    public RateLimitConfig(ProxyManager<String> buckets) {
        this.proxyManager = buckets;
    }

    public Bucket resolveBucket(String key) {
        return proxyManager.builder().build(key, getConfigSupplier());
    }

    public ArrayList<Bucket> resolveBucket(BucketConfigurationDTO bucketConfigurationDTO) {
        String path = bucketConfigurationDTO.getPath();
        ArrayList<Bucket> bucketList = new ArrayList<>();
        List<String> storeIds = bucketConfigurationDTO.getStoreIds();

        for (String storeId : storeIds) {
            String key = storeId + "--" + path;
            BucketConfiguration newBucketConfiguration = createNewBucketConfiguration(bucketConfigurationDTO.getBandwidth());
            Bucket newBucket = proxyManager.builder().build(key, getConfigSupplier());
            newBucket.replaceConfiguration(newBucketConfiguration, TokensInheritanceStrategy.AS_IS);
            bucketList.add(newBucket);
        }
        return bucketList;
    }

    private Supplier<BucketConfiguration> getConfigSupplier() {
        Refill refill = Refill.intervally(10, Duration.ofMinutes(1));
        Bandwidth limit = Bandwidth.classic(100, refill);

        return () -> (BucketConfiguration.builder()
                .addLimit(limit)
                .build());
    }

    private BucketConfiguration createNewBucketConfiguration(Bandwidth bandwidth) {
        Refill refill = Refill.intervally(bandwidth.getRefillTokens(), Duration.ofMinutes(1));
        Bandwidth limit = Bandwidth.classic(bandwidth.getCapacity(), refill);
        List<Bandwidth> bandwidthList = Arrays.asList(limit);
        return new BucketConfiguration(bandwidthList);
    }
```

The DTO class:

```
public class BucketConfigurationDTO {
    private List<String> storeIds;
    private String path;
    private Bandwidth bandwidth;
}
```

Now, when I execute:

```
BucketConfigurationDTO bucketConfigurationDTO = new BucketConfigurationDTO();
        Refill refill = Refill.of(5, Duration.ofMinutes(1));
        Bandwidth bandwidth = Bandwidth.classic(50, refill);
        bucketConfigurationDTO.setStoreIds(Arrays.asList("store13"));
        bucketConfigurationDTO.setPath("/example/path");
        bucketConfigurationDTO.setBandwidth(bandwidth);

for (int i = 0; i < 100; i++) {
     rateLimitService.allowOrLimit("store13", "/example/path");
}
```

I am supposed to have 0 tokens until it renews some. That works great

The problem is that when I run configureBucket(bucketConfigurationDTO), the bucket returned from rateLimitConfig.resolveBucket(bucketConfig) is the same bucket with the default config (100 limit for example).
Thus I know the replacement failed.

If I understand correctly, the line:
return proxyManager.builder().build(key, getConfigSupplier());
Should create a new bucket only if there is no existing bucket with the same key, am I right?
If so, I don't see where the problem can come from.

**Reply by vladimir-bukhtoyarov** [reply]
> If I understand correctly, the line:
return proxyManager.builder().build(key, getConfigSupplier());
Should create a new bucket only if there is no existing bucket with the same key, am I right?
If so, I don't see where the problem can come from.

The problem can be in expiration, if bucket expired after reconfiguration then it will be recreated with hardcoded params. What technology do you use? In which way do you initialize ProxyManager,

**Reply by AmitLustiger** [reply]
```
@Configuration
public class RedisConfig {

    @Bean
    public Config config() {
        Config config = new Config();
        config.useSingleServer().setAddress("redis://localhost:6379");
        return config;
    }

    @Bean
    public CacheManager cacheManager(Config config) {
        CacheManager cacheManager = Caching.getCachingProvider("org.redisson.jcache.JCachingProvider").getCacheManager();
        cacheManager.createCache("cache", RedissonConfiguration.fromConfig(config));
        return cacheManager;
    }

    @Bean
    ProxyManager<String> proxyManager(CacheManager cacheManager) {
        return new JCacheProxyManager<>(cacheManager.getCache("cache"));
    }

    @Bean
    @Primary
    public SyncCacheResolver bucket4jCacheResolver(CacheManager cacheManager) {
        return new JCacheCacheResolver(cacheManager);
    }
```

**What is the expiration of a bucket?**
I just create a bucket and 2 sec after it try to reconfigure it.
I run through a test.

**Reply by vladimir-bukhtoyarov** [reply]
I am not sure that Redisson over JCache is good choice for Bucket4j. I have got many claims before that this way is not working. Firstly, I would recommend to run this compatibility test https://bucket4j.com/8.7.0/toc.html#verification-of-compatibility-with-a-particular-jcache-provider-is-your-responsibility

I recommend to use direct integration with Redisson https://bucket4j.com/8.7.0/toc.html#example-of-bucket-instantiation-via-redissonbasedproxymanager

**Reply by AmitLustiger** [reply]
Eventually, I used LettuceBasedProxyManager and it still returns a bucket with 100 tokens capacity even though I reconfigure a bucket with the same ID with other Bandwidth settings (capacity 50 for example).


I think something in the replaceConfiguration method not working as supposed to.

`
Bucket bucket =  proxyManager.builder().build(key, getConfigSupplier());
        Refill refill = Refill.intervally(20, Duration.ofMinutes(1));
        Bandwidth limit = Bandwidth.classic(40, refill);
        bucket.replaceConfiguration(createNewBucketConfiguration(limit), TokensInheritanceStrategy.PROPORTIONALLY);
        return bucket;
`

Where `private Supplier<BucketConfiguration> getConfigSupplier() {
        Refill refill = Refill.intervally(10, Duration.ofMinutes(1));
        Bandwidth limit = Bandwidth.classic(100, refill);

        return () -> (BucketConfiguration.builder()
                .addLimit(limit)
                .build());
    }`

and 

`private BucketConfiguration createNewBucketConfiguration(Bandwidth bandwidth) {
        Refill refill = Refill.intervally(bandwidth.getRefillTokens(), Duration.ofMinutes(1));
        Bandwidth limit = Bandwidth.classic(bandwidth.getCapacity(), refill);
        List<Bandwidth> bandwidthList = Arrays.asList(limit);
        ConfigurationBuilder newConfiguration = BucketConfiguration.builder()
                .addLimit(limit);
        return newConfiguration.build();
    }`







Also, it's not very clear how the versions thing works using withImplicitConfigurationReplacement.
If I get it right, you need to create the bucket using it, giving it some version, and if you want to reconfigure it, you create another bucket with the same key but with a **lower** version number.

If so, let's say I want to be able to reconfigure a bucket an unknown number of times...
how should I initialize it? because if I use a specific number and each configuration I subtract 1 from it, eventually I reach zero.
Another thing is, do I have any access to the version of a bucket if I only hold its key? Otherwise, I can't know which version to give it when reconfiguring it. (I can save it in a data structure of course, but I think it's more useful to have a method that returns it by the bucket key).

**Additional notes**
In your example, you write:
`org.redisson.command.CommandExecutor commandExecutor = ...;
RedissonBasedProxyManager<String> proxyManager = RedissonBasedProxyManager.builderFor(commandExecutor)
    `

But you don't demonstrate how to initialize the commandExecutor.

**I investigated it and my conclusion is:**

Given that Redisson does not expose CommandAsyncExecutor directly, and it is an internal component of Redisson, initializing RedissonBasedProxyManager directly can be complex. 
If RedissonBasedProxyManager is a requirement, it might involve delving into the internals of Redisson or potentially modifying its source, which is not recommended due to maintenance and upgrade challenges.

I guess many others will struggle with it.

Thank you

**Reply by vladimir-bukhtoyarov** [reply]
> Eventually, I used LettuceBasedProxyManager and it still returns a bucket with 100 tokens capacity even though I reconfigure a bucket with the same ID with other Bandwidth settings (capacity 50 for example).

It looks that we waste time. Please write unit test that demonstrates the problem, use this package https://github.com/bucket4j/bucket4j/tree/master/bucket4j-redis/src/main/java/io/github/bucket4j/redis/redisson/cas

**Reply by AmitLustiger** [reply]
I use Lettuce now, it does not matter to me...
Let's put it aside.

I think something in the replaceConfiguration method is not working as supposed to.

```
Bucket bucket =  proxyManager.builder().build(key, getConfigSupplier()); 
Refill refill = Refill.intervally(20, Duration.ofMinutes(1)); 
Bandwidth limit = Bandwidth.classic(40, refill); 
bucket.replaceConfiguration(createNewBucketConfiguration(limit), TokensInheritanceStrategy.PROPORTIONALLY); 
return bucket;
```

Where

```
private Supplier getConfigSupplier() {
Refill refill = Refill.intervally(10, Duration.ofMinutes(1));
Bandwidth limit = Bandwidth.classic(100, refill);

    return () -> (BucketConfiguration.builder()
            .addLimit(limit)
            .build());
}
```

And

```
private BucketConfiguration createNewBucketConfiguration(Bandwidth bandwidth) { 
Refill refill = Refill.intervally(bandwidth.getRefillTokens(), Duration.ofMinutes(1)); 
Bandwidth limit = Bandwidth.classic(bandwidth.getCapacity(), refill); 
List<Bandwidth> bandwidthList = Arrays.asList(limit); 
ConfigurationBuilder newConfiguration = BucketConfiguration.builder() .addLimit(limit); 
return newConfiguration.build(); 
}
```

Am I using the replaceConfiguration right?
